File: Leetcode/463.cpp
class Solution {
public:
    int islandPerimeter(vector<vector<int>>& grid) {
        constexpr int dx[] = {1, -1, 0, 0};
        constexpr int dy[] = {0, 0, -1, 1};
        constexpr int STEPS = 4;
        int counter = 0;
        int n = (int)grid.size();
        int m = (int)grid.front().size();
        int new_i, new_j;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (grid[i][j]) {
                    for (int step = 0; step < STEPS; ++step) {
                        new_i = i + dx[step];
                        new_j = j + dy[step];
                        if (new_i < 0 || new_i >= n || new_j < 0 || new_j >= m || !grid[new_i][new_j]) {
                            ++counter;
                        }
                    }
                }
            }
        }
        return counter;
    }
};

File: Leetcode/463/463.go
const STEPS = 4

var (
	dx = [STEPS]int{0,0,-1,1}
	dy = [STEPS]int{-1,1,0,0}
	vis [][]bool
)

func dfs(i, j, n, m int, grid [][]int) {
	var newI, newJ int
	for step := 0; step < STEPS; step++ {
		newI = i + dx[step]
		newJ = j + dy[step]
		if newI >= 0 && newI < n && newJ >= 0 && newJ < m && grid[newI][newJ] == 1 && !vis[newI][newJ] {
			vis[newI][newJ] = true
			dfs(newI, newJ, n, m, grid)
		}
	}
}

func islandPerimeter(grid [][]int) int {
	n := len(grid)
	m := len(grid[0])
	vis = nil
	vis = make([][]bool, n)
	for i := 0; i < n; i++ {
		vis[i] = make([]bool, m)
	}
	flag := false
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == 1 {
				vis[i][j] = true
				dfs(i, j, n, m, grid)
				flag = true
				break
			}
		}
		if flag {
			break
		}
	}
	ans := 0
	var newI, newJ int
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if vis[i][j] {
				for step := 0; step < STEPS; step++ {
					newI = i + dx[step]
					newJ = j + dy[step]
					if newI >= 0 && newI < n && newJ >= 0 && newJ < m {
						if grid[newI][newJ] == 0 {
							ans++
						}
					} else {
						ans++
					}
				}
			}
		}
	}
	return ans
}